src/rnp/rnpcfg.c

```diff
@@ -0,0 +1,502 @@
+{
+    int         passfd;
+    const char *stream;
+
+    /* enabling core dumps if user wants this */
+    if (rnp_cfg_getint(cfg, CFG_COREDUMPS)) {
+        params->enable_coredumps = 1;
+    }
+
+    /* checking if password input was specified */
+    if ((passfd = rnp_cfg_getint(cfg, CFG_PASSFD))) {
+        params->passfd = passfd;
+    }
+
+    /* stdout/stderr and results redirection */
+    if ((stream = rnp_cfg_get(cfg, CFG_IO_OUTS))) {
+        params->outs = stream;
+    }
+
+    if ((stream = rnp_cfg_get(cfg, CFG_IO_ERRS))) {
+        params->errs = stream;
+    }
+
+    if ((stream = rnp_cfg_get(cfg, CFG_IO_RESS))) {
+        params->ress = stream;
+    }
+
+    /* detecting keystore pathes and format */
+    if (!rnp_cfg_get_ks_info(cfg, params))
+        return RNP_FAIL;
+
+    /* default key/userid */
+    if (!rnp_cfg_get_defkey(cfg, params))
+        return RNP_FAIL;
+
+    return RNP_OK;
+}
+
+/* find the value name in the rnp_cfg */
+static int
+rnp_cfg_find(rnp_cfg_t *cfg, const char *key)
+{
+    unsigned i;
+
+    for (i = 0; i < cfg->count && strcmp(cfg->keys[i], key) != 0; i++)
+        ;
+    return (i == cfg->count) ? -1 : (int) i;
+}
+
+/* resize keys/vals arrays to the new size. Only expanding is supported */
+static int
+rnp_cfg_resize(rnp_cfg_t *cfg, unsigned newsize)
+{
+    char **temp;
+
+    if (cfg->size == 0) {
+        /* only get here first time around */
+        cfg->keys = calloc(sizeof(char *), newsize);
+        cfg->vals = calloc(sizeof(char *), newsize);
+
+        if ((cfg->keys == NULL) || (cfg->keys == NULL)) {
+            (void) fprintf(stderr, "rnp_cfg_resize: bad alloc\n");
+            return RNP_FAIL;
+        }
+        cfg->size = newsize;
+    } else if (cfg->count == cfg->size) {
+        /* only uses 'needed' when filled array */
+        temp = realloc(cfg->keys, sizeof(char *) * newsize);
+        if (temp == NULL) {
+            (void) fprintf(stderr, "rnp_cfg_resize: bad realloc\n");
+            return RNP_FAIL;
+        }
+        cfg->keys = temp;
+
+        temp = realloc(cfg->vals, sizeof(char *) * newsize);
+        if (temp == NULL) {
+            (void) fprintf(stderr, "rnp_cfg_resize: bad realloc\n");
+            return RNP_FAIL;
+        }
+        cfg->vals = temp;
+        cfg->size = newsize;
+    }
+
+    return RNP_OK;
+}
+
+/* set val for the key in config. key and val are duplicated */
+int
+rnp_cfg_set(rnp_cfg_t *cfg, const char *key, const char *val)
+{
+    char *newval = NULL;
+    char *newkey;
+    int   i;
+
+    /* protect against the case where 'value' is rnp->value[i] */
+    if (val != NULL) {
+        newval = rnp_strdup(val);
+        if (newval == NULL) {
+            (void) fprintf(stderr, "rnp_cfg_set: bad alloc\n");
+            return RNP_FAIL;
+        }
+    }
+
+    if ((i = rnp_cfg_find(cfg, key)) < 0) {
+        /* add the element to the array */
+        if (rnp_cfg_resize(cfg, cfg->size + 15)) {
+            newkey = rnp_strdup(key);
```

Comment:
The key must be null terminated, I added this to the function description now.
Also it is used internally so our task to ensure that all calls are made according to this requirement.